server/controllers/chartsController/helpers.js

```diff
@@ -303,3 +333,74 @@ export const calculateSubjectVariableDayCount = (
 
   return subjectsDayDataAndVariable.length
 }
+
+export const buildFacetForFilters = ({
+  isSexAtBirthFilterActive,
+  isInclusionCriteriaFilterActive,
+  userAccess,
+}) => {
+  const facetForFilters = {}
+
+  if (isSexAtBirthFilterActive) {
+    facetForFilters.socioDemographics = [
+      {
+        $match: {
+          assessment: SOCIODEMOGRAPHICS_FORM,
+          study: { $in: userAccess, $nin: STUDIES_TO_OMIT },
+        },
+      },
+      {
+        $project: {
+          ...INDIVIDUAL_FILTERS_MONGO_PROJECTION,
+          socioDemographicsCollection: MONGO_COLLECTION_STRING,
+        },
+      },
+    ]
+  }
+  if (isInclusionCriteriaFilterActive) {
+    facetForFilters.inclusionCriteria = [
+      {
+        $match: {
+          assessment: INCLUSION_EXCLUSION_CRITERIA_FORM,
+          study: { $in: userAccess, $nin: STUDIES_TO_OMIT },
+        },
+      },
+      {
+        $project: {
+          ...INDIVIDUAL_FILTERS_MONGO_PROJECTION,
+          inclusionCriteriaCollection: MONGO_COLLECTION_STRING,
+        },
+      },
+    ]
+  }
+
+  return facetForFilters
+}
+
+export const intersectSubjectsFromFilters = (filters) => {
+  const intersectedSubjects = new Map()
+  const [filterListA, ...rest] = Object.values(filters).map((filter) => {
+    const filterMap = new Map()
+    filter.forEach((subjectData) => {
+      filterMap.set(subjectData.subject, { ...subjectData })
+    })
+
+    return filterMap
+  })
```

Comment:
this is iterating through an array of objects to turn them into maps. I'm not sure that computation is helpful, what do you think about using `const [filterListA, ...rest] = Object.values(filters)` directly instead?

Reply:
oh wait, this seems to be doing some re-organization as well. Is this grouping the subjects together?

Reply:
I'm sorry, maybe I missed something from our conversation yesterday and the pseudo code.

```suggestion
const intersectedSubjects = (filters) => {
  const sets = filters.map(filter => new Set([...Object.keys(filter)]))

  // intersect the sets

  // return an array of subjects that are in every filter
}
```

Reply:
This is setting the subject id as the key and the toc query results for the assessments of the filters requested, so that in the next step when we intersect the subjects we can assert that they're there using the `has` method

Reply:
sorry, I don't understand your comment. It seems like what we're doing here is transforming the filters of shape

```
 { 
   filterA: [
     {
        subject,
        filterACollection,
        study
    }
  ],
  filterB: [
      {
        subject,
        filterBCollection,
        study
      }
  ]
}
```

to an array of maps of shape

```js
{
  [subject]: { subject, filterACollection, study }
}[]
```

which makes sense, though I think it would be awesome to have the shapes the same if it makes sense to do so.

```
 { 
   filterA: [
     {
        subject,
        filter,
        collection,
        study
    }
  ],
  filterB: [
      {
        subject,
        filter,
        collection,
        study
      }
  ]
}
```

becomes

```js
{
  [subject]: { subject, filter, collection, study }
}[]
```

Reply:
https://github.com/AMP-SCZ/dpdash/pull/280/commits/5cac1a4ad5d269fb70017befbad1aca505307a18